binary-data-formats.qmd

```diff
@@ -0,0 +1,236 @@
+---
+freeze: true
+---
+
+# Binary data formats {#sec-binary-data}
+
+```{r setup}
+#| results: "asis"
+#| echo: false
+source("_common.R")
+status("drafting")
+```
+
+## Introduction
+
+### Prerequisites
+
+This chapter focuses on large data sets, and binary file formats that are well suited to working with them.
+We'll use a data set of item checkouts from Seattle public libraries, available online at [data.seattle.gov/Community/Checkouts-by-Title/tmmm-ytt6](https://data.seattle.gov/Community/Checkouts-by-Title/tmmm-ytt6).
+The data is distributed as a single 9GB csv file, so it may take some time to download.
+For the purposes of this chapter we'll assume that you have a local copy of the file at `data/seattle-library-checkout.csv`.
+
+```{r load-packages}
+#| message: false
+#| warning: false
+library(tidyverse)
+library(arrow)
+library(duckdb)
+library(dbplyr)
+```
+
+## The arrow package
+
+In this chapter we'll rely on Apache Arrow ([arrow.apache.org](https://arrow.apache.org/)), a multi-language toolbox designed for efficient analysis and transport of large data sets.
+The [arrow package](https://arrow.apache.org/docs/r/) provides a standard way to use Apache Arrow in R, and allows you to analyze larger-than-memory datasets using familiar dplyr syntax.
+As an additional benefit, arrow is extremely fast: you'll see some examples later in the chapter.
+
+### Opening datasets
+
+Let's start by taking a look at the Seattle library checkout data.
+At 9GB in size, the csv file is large enough that we probably don't want to load the whole thing into memory: a good rule of thumb is that you usually want at least twice as much memory as the size of the data.
+Instead of using `read_csv()` we'll use the `open_dataset()` function from the arrow package:
+
+```{r open-dataset}
+seattle_lib <- open_dataset(
+  sources = "data/seattle-library-checkouts.csv", 
+  format = "csv"
+)
+```
+
+When this code is run, `open_dataset()` scans the first part of the file: it reads enough rows to work out the structure of the data set structure and creates the `seattle_lib` object that stores metadata.
+The data remain on-disk, and are not loaded into memory.
+Nevertheless, we are able to call dplyr functions to interact with the data:
+
+```{r glimpse-data}
+#| cache: true
+glimpse(seattle_lib)
+```
+
+The first line in the output tell you that `seattle_lib` is stored locally on-disk as a single csv file.
+The second line indicates that the data are stored as a tabular data set with 41 million rows and 12 columns.
+
+### Partitioning datasets
+
+When data sets become large it is often useful to split them across many files rather than storing everything in a single large file.
+When this structuring is done intelligently, the multi-file strategy can lead to significant improvements in performance because some analyses will only require a subset of the files to be read.
```

Comment:
And you your computer can work on multiple pieces in parallel? Or is that true if you use parquet even without partitioning?

Reply:
I am pretty sure this is true for parquet data even without partitioning because the Dataset API can leverage the internal row grouping within the parquet file. I can't verify the difference using a dplyr (because honestly I haven't quite worked out all the details of how the dplyr code translates!), but here's an example using the low level Scanner API:

``` r
library(arrow, warn.conflicts = FALSE)
seattle_lib <- open_dataset("data/seattle-library-checkouts.parquet")

multi_threaded_scan <- Scanner$create(
  dataset = seattle_lib,
  filter = Expression$field_ref("CheckoutYear") == 2021,
  use_threads = TRUE
)
system.time(multi_threaded_scan$ToTable())
#>    user  system elapsed 
#>  22.341   6.826   7.924

single_threaded_scan <- Scanner$create(
  dataset = seattle_lib,
  filter = Expression$field_ref("CheckoutYear") == 2021,
  use_threads = FALSE
)
system.time(single_threaded_scan$ToTable())
#>    user  system elapsed 
#>  16.699   5.840  20.603
```

<sup>Created on 2022-12-01 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Nice, thanks!